crates/hir-ty/src/method_resolution.rs

```diff
@@ -1478,26 +1478,38 @@ fn is_valid_fn_candidate(
             // We need to consider the bounds on the impl to distinguish functions of the same name
             // for a type.
             let predicates = db.generic_predicates(impl_id.into());
-            let valid = predicates
-                .iter()
-                .map(|predicate| {
-                    let (p, b) = predicate
-                        .clone()
-                        .substitute(Interner, &impl_subst)
-                        // Skipping the inner binders is ok, as we don't handle quantified where
-                        // clauses yet.
-                        .into_value_and_skipped_binders();
-                    stdx::always!(b.len(Interner) == 0);
-                    p
-                })
-                // It's ok to get ambiguity here, as we may not have enough information to prove
-                // obligations. We'll check if the user is calling the selected method properly
-                // later anyway.
-                .all(|p| table.try_obligation(p.cast(Interner)).is_some());
-            match valid {
-                true => IsValidCandidate::Yes,
-                false => IsValidCandidate::No,
+            let mut alias = Vec::new();
+            let mut other_predicate = Vec::new();
+
+            for predicate in predicates.iter() {
+                let (p, b) = predicate
+                    .clone()
+                    .substitute(Interner, &impl_subst)
+                    // Skipping the inner binders is ok, as we don't handle quantified where
+                    // clauses yet.
+                    .into_value_and_skipped_binders();
+                stdx::always!(b.len(Interner) == 0);
+
+                if let WhereClause::AliasEq(_) = p {
+                    alias.push(p);
+                } else {
+                    other_predicate.push(p);
+                }
+            }
+
+            for p in alias {
+                if !table.try_resolve_alias(p.cast(Interner)) {
```

Comment:
I think I would prefer if this applied the solution here, instead of hiding that inside `try_resolve_alias`. Also, is there a reason to only do this for aliases? What does rustc do here?

Reply:
> I think I would prefer if this applied the solution here

I'd also like to, but `apply_solution` is not public here.

> Also, is there a reason to only do this for aliases?

There are four kinds of `WhereClause`, and I believe only `AliasEq` would generate solution. But it does no harm to check everything for `Solution`.

Reply:
I think making `apply_solution` `pub(crate)` would be fine.

> There are four kinds of WhereClause, and I believe only AliasEq would generate solution. But it does no harm to check everything for Solution.

Yeah, I think it's better to just do it for everything.